### DISCUSSION TITLE
Markdown to Typst: Typographic substitutions

### DISCUSSION THREAD

**Discussion by johnmapeson** [jgm/pandoc#10129] [Q&A]
Test file `input.md`:

```
= Lorem ipsum

Dolor sit --- amet. Dolor "sit" amet.
```

- If I convert it to Typst using `pandoc -o output.typ input.md`, the output is

  ```
  = Lorem ipsum
  <lorem-ipdum>
  Dolor sit — amet. Dolor "sit" amet.
  ```
  
  (`---` is replaced with em dash.)

- If I convert it to Typst using `pandoc -f markdown-smart -o output.typ input.md`, the output is

  ```
  = Lorem ipsum
  <lorem-ipdum>
  Dolor sit --- amet. Dolor \"sit\" amet.
  ```

Questions:

- How is it possible to convert from Markdown to Typst so that the output will be

  ```
  = Lorem ipsum
  <lorem-ipdum>
  Dolor sit — amet. Dolor “sit” amet.
  ```

  (`---` is replaced with em dash. Straight quotation marks are replaced with curly ones.)

- How is it possible to convert from Markdown to Typst so that the output will be

  ```
  = Lorem ipsum
  <lorem-ipdum>
  Dolor sit --- amet. Dolor "sit" amet.
  ```

  (Quotation marks are straight and are not escaped. `---` is not replaced with em dash.)

**Comment by jgm** [comment]
First, you probably want `#` instead of `=` for the markdown heading.

Pandoc parses this as
```hs
[ Header
    1
    ( "lorem-ipsum" , [] , [] )
    [ Str "Lorem" , Space , Str "ipsum" ]
, Para
    [ Str "Dolor"
    , Space
    , Str "sit"
    , Space
    , Str "\8212"
    , Space
    , Str "amet."
    , Space
    , Str "Dolor"
    , Space
    , Quoted DoubleQuote [ Str "sit" ]
    , Space
    , Str "amet."
    ]
]
```
but the Quoted element is rendered to typst using straight quotes, because typst will interpret them as if they're curly.  (You'll get curly quotes after you compile with typst.)

**Comment by johnmapeson** [comment]
> First, you probably want `#` instead of `=` for the markdown heading.

Yes, sorry. My bad, I was in a hurry.

Not sure I understand the rest part...

**Comment by cskeeters** [comment]
# Quotes

Here, I get why you'd want Typst to handle.  The user may have a custom template with `#smartquote` configured.

| Smart     | Markdown     | Typst         | Comment                       |
| --------- | ------------ | ------------- | ----------------------------- |
| +smart    | I said "no"  | I said "no".  | Nice, let Typst handle        |
| +smart    | I said 'no'  | I said 'no'.  | Nice, let Typst handle        |
| -smart    | I said "no"  | I said "no".  | Fine, let template handle     |
| -smart    | I said 'no'  | I said 'no'.  | Fine, let template handle     |


# Apostrophes

When smart is on, why not let Typst handle apostrophes?

| Smart     | Markdown      | Typst         | Comment                          |
| --------- | ------------  | ------------- | -------------------------------- |
| +smart    | Don't do it   | Don’t do it   | **Why not let Typst handle?**    |
| +smart    | Don\\'t do it | Don\\'t do it | Makes sense                      |
| -smart    | Don't do it   | Don\\'t do it | Why not let the template handle? |
| -smart    | Don\\'t do it | Don\\'t do it | Makes sense                      |

**Reply by jgm** [reply]
Note: `smart` can be attached to both markdown and typst.  So there are actually many more combinations than you have represented:  `-f markdown+smart -t typst+smart`, `-f markdown+smart -t typst-smart`, `-f markdown-smart -t typst+smart`, `-f markdown-smart -t typst-smart`.

**Reply by cskeeters** [reply]
Oh, dang.  I even read that extensions can be applied to input and output formats, but it didn't occur to me how to configure it in the output format until now.  Of course, now that you mention it, it seems obvious.

That said, it seems that the Typst format doesn't support the *smart* extension.  https://pandoc.org/MANUAL.html#extension-smart

```
The extension smart is not supported for typst.
Use --list-extensions=typst to list supported extensions.
```

The above table stands assuming `-f markdown(+-)smart` and `-t typst`.

It seems that if `-f markdown+smart` is specified the apostrophe is converted to a right single quotation mark (U+2019) as it's converted to the AST.

    [ Para
        [ Str "Don\8217t" , Space , Str "do" , Space , Str "it." ]
    ]

I want typst to convert to smartquotes at `typst compile`-time, so I guest I'd use `-f markdown-smart` and then `-t typst` (typst-smart when supported).  In this case, the stock Typst writer(?) escapes the apostrophe.  It would be nice if it didn't do that, and allowed the template author to set it with `#smartquote`.

**Reply by jgm** [reply]
Oh, I forgot that typst doesn't enable the `smart` extension.  (That makes sense, because "smart" treatment of these characters is built into typst and can't be disabled, as far as I know.)

`-f markdown-smart` will parse a `'` character as a literal `'`, and render it appropriately (hence using the escape in typst to avoid the typographic substitution).

Here's a possible change to the typst writer that is suggested by your post:  we could simply render the `’` character as `'`.  That would be consistent with how quotes are being handled currently.  Feel free to create an issue making this suggestion.  Perhaps we should do something similar with ellipses and dashes, if we don't already.

**Reply by cskeeters** [reply]
> can't be disabled, as far as I know

I tested it, and it can, as follows:

```
#set smartquote(enabled: false)
```

I created an issue: <https://github.com/jgm/pandoc/issues/10271>

**Reply by thjbdvlt** [reply]
Hello!

Typst `smartquotes` is actually not very functional. For french, it's unusable because when `"` is preceded by a punctuation, it will not be converted into a quote:

```md
this "will parsed/processed successfully".
but this won't: l'"autre" (very common in french)
```

output using Typst writer (with `smartquotes`):

```txt
this « will be parsed/processed successfully ».
but this won't: l'"autre" (very common in french)
```

as Pandoc do parse this successfully, maybe it would make sense to use the Typst `#quote` function?

like this (so it can be configurated, for example, in the typst template):

```typst
#show quote: set text(color: red)
this #quote([will be parsed/processed successfully]).
but this wont': l'#quote([autre]) (very common in french).
```

of course, it's not a huge problem, one could just use pandoc-quotes lua filter, but it has the caveats of losing the semantic `Quote` parsed elements created by Pandoc (that could be used in Typst).

(i hope i am not too unclear)

**Reply by jgm** [reply]
Interesting suggestion.  Looks like `#quote` can be affected by e.g. `#quote(block: true)` which would completely change the semantics -- that's why I'm reluctant.

**Reply by thjbdvlt** [reply]
Maybe it could somehow be an option, like the `--html-q-tags` (pretty similar case?). Of course it better be explicitely called as `#quote(block: false, quotes: true, attribution: none, [content])` (still let the user able to configure them independenly of blockquotes, using `#set quote.where(block: false)...`).

**Reply by jgm** [reply]
Yes, we could emit that but it makes the output ugly.  So, it's probably something we'd only want to do under an option like `--typst-quote-style` or something.

**Reply by thjbdvlt** [reply]
Absolutely ugly, yes! I was actually thinking about Typst as `--pdf-engine`; you're right that it would make the `.typ` unreadable.

(An alias function may be used, like `#q[...]` but it would probably be weird to define aliases function there.)

**Reply by jgm** [reply]
That's true, `#q` is not a bad idea, and it could be defined in our default template.  It makes the typst less portable though.

**Reply by cskeeters** [reply]
I've been testing main since you put in [the fix](https://github.com/jgm/pandoc/commit/bdb1172385422d46243e6b5ede31fb6054fade5f#diff-b6ab37e5fd63cad3e38f5fb7d41b481ccfbbedf112595d592f012b110992cf80R31) for me.  I love it!  Thank you.

I have concluded that you can use it two ways.  Re-explaining helps me flush out the detail...

Say you have `smart.md`:

```markdown
First -- Last

First --- Last

I said "no"

Don't do it
````

NOTE: This is valid Typst code and will utilize smart quotes on compile, so I want the output to be the same as the input, disregarding stuff related to making a standalone document.

If one runs `pandoc smart.md -t typst`, they are now effectively running `pandoc -f markdown+smart smart.md -t typst+smart` and the output will be unchanged from the markdown source.  **Yay!**  With or without `-s` (standalone) the dashes and quotes will get converted by `typst compile`.  When run like this, the markdown reader does converts dashes and strait quotes into curly versions in the AST, but typst+smart converts them back before being output.

(Made with `pandoc smart.md -t native`)
```
[ Para
    [ Str "First" , Space , Str "\8211" , Space , Str "Last" ]
, Para
    [ Str "First" , Space , Str "\8212" , Space , Str "Last" ]
, Para
    [ Str "I"
    , Space
    , Str "said"
    , Space
    , Quoted DoubleQuote [ Str "no" ]
    ]
, Para
    [ Str "Don\8217t" , Space , Str "do" , Space , Str "it" ]
]
```
NOTE: \8211 used for en-dash, \8212 used for em-dash, Quoted DoubleQuote and \8217 for apostrophe.

If one runs `pandoc -f markdown-smart smart.md -t typst-smart`, the same output is generated without `-s`, but the native AST does not have unicode representations in the AST.

(Made with `pandoc -f markdown-smart smart.md -t native`)
```
[ Para
    [ Str "First" , Space , Str "--" , Space , Str "Last" ]
, Para
    [ Str "First" , Space , Str "---" , Space , Str "Last" ]
, Para
    [ Str "I" , Space , Str "said" , Space , Str "\"no\"" ]
, Para [ Str "Don't" , Space , Str "do" , Space , Str "it" ]
]
```

NOTE: Mostly no translation.
NOTE: Escaped double strait quotes. (? -- See question below)

The new default template for typst will output `#set smartquote(enabled: false)` when `typst-smart` is specified which makes quotes strait in the compiled PDF when run translated with `-s` and compiled with `typst compile`.

It appears a second way to achieve curly quotes it to run `pandoc -s -f markdown-smart smart.md -t typst-smart --template <custom>` where the custom template has the [if block](https://github.com/jgm/pandoc/blob/32c1a31f61b012aaae784138fb2a049ccdd496ad/data/templates/default.typst#L31) with `#set smartquote(enabled: false)` removed.  Since I have my own templates already, I just have to not add it.  Both ways work with the latter option being closer to what I intuitively want to happen.

**Question**: Why are the double quotes escaped in the AST for `markdown-smart` input?

**Reply by jgm** [reply]
This escaping only relates to the Haskell representation (because " is used for strings, you need `\"` for a literal quote inside a string). Don't worry, you won't get a backslash in rendered output.

**Reply by cskeeters** [reply]
Ah, of course.  Thx for the reply.
